File: LeetCode/Problems/C++/51. N-Queens.cpp
class Solution {
public:
    bool checkValid(vector<int> queens, int i, int n){
        //check vertical
        for(auto y:queens){
            if(i==y) return false;
        }
        
        //check diagonal
        int offset = 1;
        for(int j=queens.size()-1;j>=0;j--){
            if((i+offset < n && queens[j] == i+offset) || (i-offset >=0 && queens[j] == i-offset))
                return false;
            offset++;
        }

        return true;
    }
    
    
    void backtracking(vector<vector<string>>& result, int row, int n, vector<int> queens) {
        if(queens.size() == n){
            //construct answer
            vector<string> found_answer;
            string ans = "";
            for(int i =0;i<n;i++){
                ans = "";
                for(int j=0;j<n;j++){
                    if(j==queens[i]) ans+='Q';
                    else ans+='.';
                }
                found_answer.push_back(ans);
            }
            result.push_back(found_answer);
            return;
        }
        
        
        for(int i=0;i<n;i++){
            if(checkValid(queens, i, n)){
                //place queens
                queens.push_back(i);
                //backtracking
                backtracking(result, row+1, n, queens);
                //reset queens
                queens.pop_back();
            }
                
        }
        
        
    }
    vector<vector<string>> solveNQueens(int n) {
        if(!n) return {};
        
        vector<vector<string>> result;
        
        backtracking(result, 0, n, {});
        
        return result;
    }
};

File: LeetCode/Problems/Python/51. N-Queens.py
class Solution:
    def backtracking(self, grid, row, n, result):
        if row == n:
            temp = deepcopy(grid)
            for i in range(n):
                temp[i] = "".join(temp[i])
            result.append(temp)
            return
        else:
            def checkValid(row, i):
                #check current col i
                for j in range(row-1,-1,-1):
                    if grid[j][i] == 'Q':
                        return False

                #check diagonal
                k=1
                for j in range(row-1,-1,-1):
                    if i+k < n and grid[j][i+k] == 'Q':
                        return False
                    if i-k >=0 and grid[j][i-k] == 'Q':
                        return False
                    k+=1

                return True


            for i in range(n):
                if checkValid(row,i):
                    grid[row][i] = 'Q'
                    self.backtracking(grid, row+1, n, result)
                    grid[row][i] = '.'

        
    
    def solveNQueens(self, n: int) -> List[List[str]]:
        if not n:
            return []
        
        result = []
        grid = [['.'] * n for _ in range(n)]
        
        self.backtracking(grid, 0, n, result)
        
        return result